buffer-api/src/main/java/io/netty/buffer/api/Resource.java

```diff
@@ -0,0 +1,41 @@
+package io.netty.buffer.api;
+
+/**
+ * A resource that has a life-time, and can be {@linkplain #close() closed}.
+ * Resources are initially {@linkplain #isAccessible() accessible}, but closing them makes them inaccessible.
+ */
+public interface Resource<T extends Resource<T>> extends AutoCloseable {
+    /**
+     * Send this object instance to another Thread, transferring the ownership to the recipient.
+     * <p>
+     * The object must be in a state where it can be sent, which includes at least being
+     * {@linkplain #isAccessible() accessible}.
+     * <p>
+     * When sent, this instance will immediately become inaccessible, as if by {@linkplain #close() closing} it.
+     * All attempts at accessing an object that has been sent, even if that object has not yet been received, should
+     * cause an exception to be thrown.
+     * <p>
+     * Calling {@link #close()} on an object that has been sent will have no effect, so this method is safe to call
+     * within a try-with-resources statement.
+     */
+    Send<T> send();
+
+    /**
+     * Close the resource, making it inaccessible.
+     * <p>
+     * Note, this method is not thread-safe unless otherwise specific.
+     *
+     * @throws IllegalStateException If this {@code Resource} has already been closed.
```

Comment:
I know this is ok from an API contract but I wonder if we should better follow what is stated in the javadocs:

```
Note that unlike the close method of Closeable, this close method is not required to be idempotent. In other words, calling this close method more than once may have some visible side effect, unlike Closeable.close which is required to have no effect if called more than once. However, implementers of this interface are strongly encouraged to make their close methods idempotent.

```

Reply:
I think calling `close` more than once is indicative of a bug, where there's uncertainty about the buffers life time. The `release` method on `ByteBuf` also throws if you call it on a buffer that's already been deallocated.